File: sdks/go/opspec/interpreter/call/op/params/param/validate.go
package param

import (
	"errors"
	"fmt"

	"github.com/opctl/opctl/sdks/go/model"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/array"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/boolean"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/dir"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/file"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/number"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/object"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/socket"
	"github.com/opctl/opctl/sdks/go/opspec/interpreter/call/op/params/param/str"
)

// Validate validates a value against a parameter
// note: param defaults aren't considered
func Validate(
	value *model.Value,
	param *model.Param,
) []error {
	if value == nil {
		return []error{errors.New("required")}
	}

	switch {
	case param.Array != nil:
		return array.Validate(value, param.Array.Constraints)
	case param.Boolean != nil:
		return boolean.Validate(value)
	case param.Dir != nil:
		return dir.Validate(value)
	case param.File != nil:
		return file.Validate(value)
	case param.Number != nil:
		return number.Validate(value, param.Number.Constraints)
	case param.String != nil:
		return str.Validate(value, param.String.Constraints)
	case param.Object != nil:
		return object.Validate(value, param.Object.Constraints)
	case param.Socket != nil:
		return socket.Validate(value)
	default:
		return []error{fmt.Errorf("unable to validate value: param was unexpected type %+v", param)}
	}
}

File: sdks/js/src/opspec/interpreter/opcall/params/param/validate.ts
import Value from '../../../../../model/value'
import Param from '../../../../../model/param'
import arrayValidate from './array/validate'
import booleanValidate from './boolean/validate'
import dirValidate from './dir/validate'
import fileValidate from './file/validate'
import numberValidate from './number/validate'
import objectValidate from './object/validate'
import socketValidate from './socket/validate'
import stringValidate from './string/validate'

  /**
   * Validates value against param
   * note: param defaults aren't considered
   * @param {Value} value
   * @param {Param} param
   * @returns {Array<String>}
   */
  export default function validate (
    value: Value,
    param: Param
  ) {
    if (!param) {
      throw new Error('param required')
    }

    if (param.array) {
      return arrayValidate(
        value.array!,
        param.array.constraints
      )
    }

    if (param.boolean) {
      return booleanValidate(
        value.boolean!
      )
    }

    if (param.dir) {
      return dirValidate(
        value.link
      )
    }

    if (param.file) {
      return fileValidate(
        value.link
      )
    }

    if (param.number) {
      return numberValidate(
        value.number!,
        param.number.constraints
      )
    }

    if (param.object) {
      return objectValidate(
        value.object!,
        param.object.constraints
      )
    }

    if (param.socket) {
      return socketValidate(
        value.socket!
      )
    }

    if (param.string) {
      return stringValidate(
        value.string!,
        param.string.constraints
      )
    }
  }
